pkg/pma/pma.c

```diff
@@ -0,0 +1,783 @@
+/// @param[in] grows_down  true if the mapping grows downward in memory.
+/// @param[in] pma         PMA this mapping belongs to.
+/// @param[in] len         Number of bytes at the beginning of the mapping to be
+///                        synced to the backing file. Must be a multiple of
+///                        kPageSz.
+/// @param[in] fd          Open file descriptor of the backing file.
+///
+/// @return 0   Synced changes successfully.
+/// @return -1  Failed to sync changes.
+static_ int
+sync_file_(const char *path,
+           void       *base,
+           bool        grows_down,
+           pma_t      *pma,
+           size_t      len,
+           int         fd);
+
+/// Get the total length in bytes of a PMA.
+///
+/// @param[in] pma
+static_ inline_ size_t
+total_len_(const pma_t *pma)
+{
+    assert(pma);
+    return (size_t)(pma->stack_start - pma->heap_start);
+}
+
+static_ int
+center_guard_page_(pma_t *pma)
+{
+    size_t heap_len  = 0;
+    size_t stack_len = 0;
+    if (pma->len_getter(&heap_len, &stack_len) == -1) {
+        fprintf(stderr,
+                "pma: failed to determine length of heap and stack\r\n");
+        return -1;
+    }
+
+    size_t free_len   = total_len_(pma) - (heap_len + stack_len);
+    char  *free_start = (char *)pma->heap_start + pma->heap_len;
+    void  *guard_pg   = free_start + round_up(free_len / 2, kPageSz);
+
+    if (mmap(guard_pg,
+             kPageSz,
+             PROT_NONE,
+             MAP_ANON | MAP_FIXED | MAP_PRIVATE,
+             -1,
+             0)
+        == MAP_FAILED)
+    {
+        fprintf(stderr,
+                "pma: failed to place %zu-byte guard page at %p: %s\r\n",
+                kPageSz,
+                guard_pg,
+                strerror(errno));
+        return -1;
+    }
+    set_page_status_(guard_pg, PS_MAPPED_INACCESSIBLE, pma);
+    pma->guard_pg = guard_pg;
+    return 0;
+}
+
+static_ int
+handle_page_fault_(void *fault_addr, void *user_arg)
+{
+    fault_addr = (void *)round_down((uintptr_t)fault_addr, kPageSz);
+    pma_t *pma = user_arg;
+    switch (page_status_(fault_addr, pma)) {
+        case PS_UNMAPPED:
+            if (MAP_FAILED
+                == mmap(fault_addr,
+                        kPageSz,
+                        PROT_READ | PROT_WRITE,
+                        MAP_ANON | MAP_FIXED | MAP_PRIVATE,
+                        -1,
+                        0))
+            {
+                fprintf(stderr,
+                        "pma: failed to create %zu-byte anonymous mapping at "
+                        "%p: %s\r\n",
+                        kPageSz,
+                        fault_addr,
+                        strerror(errno));
+                return 0;
+            }
+            set_page_status_(fault_addr, PS_MAPPED_DIRTY, pma);
+            break;
+        case PS_MAPPED_CLEAN:
+            if (mprotect(fault_addr, kPageSz, PROT_READ | PROT_WRITE) == -1) {
```

Comment:
Isn't it our codebase's convention to place the constant value on the left?

Reply:
Yes, but our convention is non-idiomatic and forces readers to effectively read from right to left. Since this is a separate library, I've taken the liberty of using purely idiomatic C.

Reply:
There really is no idiom, only war. For long function calls and a comparison against a tiny (textually) literal, you can see why some prefer the other. Moreover it's arguably safer